File: timesfm3/serving/credits.py
"""Server side of the credit pool: signing, redemption, nullifiers."""

from __future__ import annotations

import json
import math
import os
import secrets
import tempfile
import threading

from ..credits import CREDIT_COSTS, SERIAL_BYTES, key_id, nullifier, verify_token
from .auth import ApiKey


# -- RSA key generation in pure Python (no OpenSSL binding needed; runs once
#    and is persisted). Miller-Rabin with 40 rounds: error < 2^-80.

_SMALL_PRIMES = [p for p in range(3, 2000, 2) if all(p % q for q in range(3, int(p ** 0.5) + 1, 2))]


def _is_probable_prime(n: int, rounds: int = 40) -> bool:
    if n < 2:
        return False
    for p in _SMALL_PRIMES:
        if n % p == 0:
            return n == p
    d, r = n - 1, 0
    while d % 2 == 0:
        d //= 2
        r += 1
    for _ in range(rounds):
        a = secrets.randbelow(n - 3) + 2
        x = pow(a, d, n)
        if x in (1, n - 1):
            continue
        for _ in range(r - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True


def _random_prime(bits: int) -> int:
    while True:
        c = secrets.randbits(bits) | (1 << (bits - 1)) | 1
        if _is_probable_prime(c):
            return c


def generate_rsa(bits: int = 2048, e: int = 65537) -> dict:
    while True:
        p, q = _random_prime(bits // 2), _random_prime(bits // 2)
        if p == q:
            continue
        n = p * q
        phi = (p - 1) * (q - 1)
        if n.bit_length() == bits and math.gcd(e, phi) == 1:
            return {"n": n, "e": e, "d": pow(e, -1, phi)}

CREDIT_IDENTITY = ApiKey(key="", name="credit-pool", plan="prepaid")
DENOMINATIONS = (10, 25, 100)
DEFAULT_PRICE_PER_CREDIT_USD = 0.004  # 20% below pay-per-call
POINTS_PER_CREDIT = 256  # what a plan holder is charged per credit bought


class CreditPool:
    def __init__(self, key_file: str | None = None, ledger_file: str | None = None,
                 price_per_credit_usd: float = DEFAULT_PRICE_PER_CREDIT_USD, bits: int = 2048):
        self._lock = threading.Lock()
        self.ledger_file = ledger_file
        self.price_per_credit_usd = price_per_credit_usd
        self.ephemeral = key_file is None
        if key_file and os.path.exists(key_file):
            with open(key_file) as f:
                key = {k: int(v, 16) for k, v in json.load(f).items()}
        else:
            key = generate_rsa(bits)
            if key_file:
                d = os.path.dirname(os.path.abspath(key_file))
                os.makedirs(d, exist_ok=True)
                fd, tmp = tempfile.mkstemp(dir=d, prefix=".credits-key-")
                with os.fdopen(fd, "w") as f:
                    json.dump({k: format(v, "x") for k, v in key.items()}, f)
                os.chmod(tmp, 0o600)
                os.replace(tmp, key_file)
        self.n, self.e, self._d = key["n"], key["e"], key["d"]
        self.kid = key_id(self.n)
        self.spent: set[str] = set()
        self.issued = 0
        self.redeemed = 0
        if ledger_file and os.path.exists(ledger_file):
            with open(ledger_file) as f:
                doc = json.load(f)
            self.spent = set(doc.get("spent", []))
            self.issued = int(doc.get("issued", 0))
            self.redeemed = int(doc.get("redeemed", 0))

    def _persist(self) -> None:
        if not self.ledger_file:
            return
        d = os.path.dirname(os.path.abspath(self.ledger_file))
        fd, tmp = tempfile.mkstemp(dir=d, prefix=".credits-", suffix=".json")
        with os.fdopen(fd, "w") as f:
            json.dump({"kid": self.kid, "issued": self.issued, "redeemed": self.redeemed,
                       "spent": sorted(self.spent)}, f)
        os.replace(tmp, self.ledger_file)

    # -- public description ------------------------------------------------

    def describe(self) -> dict:
        return {
            "kid": self.kid, "n": format(self.n, "x"), "e": self.e, "bits": self.n.bit_length(),
            "scheme": "RSA-FDH blind signature (Chaum); serial = nullifier",
            "denominations": list(DENOMINATIONS),
            "price_per_credit_usd": self.price_per_credit_usd,
            "points_per_credit": POINTS_PER_CREDIT,
            "costs": dict(CREDIT_COSTS),
            "pool": {"issued": self.issued, "redeemed": self.redeemed, "outstanding": self.issued - self.redeemed},
            "ephemeral_key": self.ephemeral,
        }

    def price_usd(self, count: int) -> str:
        return f"${count * self.price_per_credit_usd:.4f}".rstrip("0").rstrip(".")

    # -- issuing -----------------------------------------------------------

    def sign_blinded(self, blinded_hex: list[str]) -> list[str]:
        out = []
        for h in blinded_hex:
            try:
                m = int(h, 16)
            except ValueError:
                raise ValueError("blinded messages must be hex integers")
            if not (0 < m < self.n):
                raise ValueError("blinded message out of range")
            out.append(format(pow(m, self._d, self.n), "x"))
        with self._lock:
            self.issued += len(out)
            self._persist()
        return out

    # -- redemption --------------------------------------------------------

    def redeem(self, header: str, cost: int) -> tuple[bool, str]:
        """Spends exactly ``cost`` valid tokens from the header; all-or-nothing."""
        tokens = [t for t in header.split(",") if t.strip()]
        if len(tokens) < cost:
            return False, f"this call costs {cost} credit(s); {len(tokens)} presented"
        serials = []
        for t in tokens[:cost]:
            serial = verify_token(t, self.n, self.e, self.kid)
            if serial is None:
                return False, "invalid credit token (bad signature or wrong key id)"
            serials.append(nullifier(serial))
        if len(set(serials)) != len(serials):
            return False, "duplicate credit token in request"
        with self._lock:
            if any(s in self.spent for s in serials):
                return False, "credit token already spent"
            self.spent.update(serials)
            self.redeemed += cost
            self._persist()
        return True, ""


def credit_cost(method: str, path: str) -> int | None:
    return CREDIT_COSTS.get(f"{method.upper()} {path}")


__all__ = ["CreditPool", "CREDIT_IDENTITY", "DENOMINATIONS", "POINTS_PER_CREDIT", "credit_cost", "SERIAL_BYTES"]

File: cloudflare/src/credits.js
/**
 * Credit pool at the edge: RFC 9474 blind signatures with @cloudflare/blindrsa-ts,
 * nullifiers in D1. Same token format and suite as the Python service, so a
 * token bought from either can be redeemed at either that shares the key.
 *
 *   secret CREDITS_PRIVATE_JWK   private RSA JWK (p,q,dp,dq,qi) -> issuing key
 *   var    CREDITS_OLD_PUBLIC_JWKS  JSON array of public JWKs still redeemable
 *   var    CREDITS_PRICE_USD     per credit (default 0.004)
 *   binding CREDITS_DB           D1 (migrations/0001_credits.sql)
 */

import { RSABSSA } from "@cloudflare/blindrsa-ts";

export const SUITE_NAME = "RSABSSA-SHA384-PSSZERO-Deterministic";
export const SERIAL_BYTES = 32;
export const DENOMINATIONS = [10, 25, 100];
export const CREDIT_COSTS = { "POST /v1/forecast": 1, "POST /v1/volatility": 1, "POST /v1/anomalies": 2, "POST /v1/backtest": 4 };
export const POINTS_PER_CREDIT = 256;

const suite = RSABSSA.SHA384.PSSZero.Deterministic();
const enc = new TextEncoder();

export const b64 = {
  encode: (u8) => btoa(String.fromCharCode(...u8)).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, ""),
  decode: (s) => Uint8Array.from(atob(s.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat((4 - (s.length % 4)) % 4)), (c) => c.charCodeAt(0)),
};

async function sha256hex(bytes) {
  return [...new Uint8Array(await crypto.subtle.digest("SHA-256", bytes))].map((b) => b.toString(16).padStart(2, "0")).join("");
}

/** kid = first 12 hex of SHA-256 over the big-endian modulus (matches Python). */
export async function keyIdFromJwk(jwk) {
  return (await sha256hex(b64.decode(jwk.n))).slice(0, 12);
}

let cached = null; // { issuing: {kid, priv, pub, jwk}, keys: Map(kid -> {pub, jwk, issuing}), price }

export async function creditPool(env) {
  if (!env.CREDITS_PRIVATE_JWK) return null;
  if (cached && cached.raw === env.CREDITS_PRIVATE_JWK) return cached;
  const priv = JSON.parse(env.CREDITS_PRIVATE_JWK);
  const pubJwk = { kty: "RSA", n: priv.n, e: priv.e, alg: "PS384", use: "sig" };
  const algo = { name: "RSA-PSS", hash: "SHA-384" };
  const privateKey = await crypto.subtle.importKey("jwk", { ...priv, alg: "PS384", ext: true }, algo, true, ["sign"]);
  const publicKey = await crypto.subtle.importKey("jwk", { ...pubJwk, ext: true }, algo, true, ["verify"]);
  const kid = await keyIdFromJwk(pubJwk);
  const keys = new Map([[kid, { kid, pub: publicKey, jwk: pubJwk, issuing: true }]]);
  for (const old of JSON.parse(env.CREDITS_OLD_PUBLIC_JWKS || "[]")) {
    const okid = await keyIdFromJwk(old);
    if (!keys.has(okid)) keys.set(okid, { kid: okid, pub: await crypto.subtle.importKey("jwk", { kty: "RSA", n: old.n, e: old.e, alg: "PS384", ext: true }, algo, true, ["verify"]), jwk: { kty: "RSA", n: old.n, e: old.e, alg: "PS384", use: "sig" }, issuing: false });
  }
  cached = { raw: env.CREDITS_PRIVATE_JWK, issuing: { kid, priv: privateKey, pub: publicKey, jwk: pubJwk }, keys, price: Number(env.CREDITS_PRICE_USD || 0.004) };
  return cached;
}

export function priceUsd(pool, count) {
  return "$" + (count * pool.price).toFixed(4).replace(/0+$/, "").replace(/\.$/, "");
}

export async function describe(pool, env) {
  let stats = { issued: 0, redeemed: 0 };
  if (env.CREDITS_DB) {
    const row = await env.CREDITS_DB.prepare("SELECT COALESCE(SUM(issued),0) AS issued, COALESCE(SUM(redeemed),0) AS redeemed FROM stats").first();
    if (row) stats = { issued: Number(row.issued), redeemed: Number(row.redeemed) };
  }
  return {
    suite: SUITE_NAME, spec: "RFC 9474", serial_bytes: SERIAL_BYTES,
    keys: [...pool.keys.values()].map((k) => ({ kid: k.kid, jwk: k.jwk, issuing: k.issuing })),
    kid: pool.issuing.kid, denominations: DENOMINATIONS, price_per_credit_usd: pool.price,
    points_per_credit: POINTS_PER_CREDIT, costs: CREDIT_COSTS,
    pool: { ...stats, outstanding: stats.issued - stats.redeemed },
    ephemeral_key: false, runs_in: "edge",
    token_format: "kid.base64url(serial).base64url(signature); header X-Credit: token[,token]",
  };
}

export async function signBlinded(pool, env, blindedList) {
  const out = [];
  for (const b of blindedList) {
    let blinded;
    try { blinded = b64.decode(String(b)); } catch { throw new Error("bad blinded message"); }
    out.push(b64.encode(await suite.blindSign(pool.issuing.priv, blinded)));
  }
  if (env.CREDITS_DB) {
    await env.CREDITS_DB.prepare("INSERT INTO stats (kid, issued, redeemed) VALUES (?1, ?2, 0) ON CONFLICT(kid) DO UPDATE SET issued = issued + ?2").bind(pool.issuing.kid, out.length).run();
  }
  return out;
}

async function verifyToken(pool, token) {
  const parts = token.trim().split(".");
  if (parts.length !== 3) return null;
  const key = pool.keys.get(parts[0]);
  if (!key) return null;
  let serial, sig;
  try { serial = b64.decode(parts[1]); sig = b64.decode(parts[2]); } catch { return null; }
  if (serial.length !== SERIAL_BYTES) return null;
  try { if (!(await suite.verify(key.pub, sig, serial))) return null; } catch { return null; }
  return { kid: key.kid, serial };
}

/** All-or-nothing redemption of `cost` tokens from the X-Credit header. */
export async function redeem(pool, env, header, cost) {
  const tokens = header.split(",").map((t) => t.trim()).filter(Boolean);
  if (tokens.length < cost) return { ok: false, why: `this call costs ${cost} credit(s); ${tokens.length} presented` };
  const nulls = [];
  for (const t of tokens.slice(0, cost)) {
    const v = await verifyToken(pool, t);
    if (!v) return { ok: false, why: "invalid credit token (bad signature or unknown key id)" };
    const nul = await sha256hex(new Uint8Array([...enc.encode("nullifier"), ...v.serial]));
    nulls.push({ nul, kid: v.kid });
  }
  if (new Set(nulls.map((n) => n.nul)).size !== nulls.length) return { ok: false, why: "duplicate credit token in request" };
  if (!env.CREDITS_DB) return { ok: false, why: "credit ledger not configured" };
  const ts = Date.now();
  // INSERT OR IGNORE per nullifier; if any was already present, roll back the batch.
  const results = await env.CREDITS_DB.batch(nulls.map((n) => env.CREDITS_DB.prepare("INSERT OR IGNORE INTO spent (nullifier, kid, ts) VALUES (?1, ?2, ?3)").bind(n.nul, n.kid, ts)));
  const inserted = results.map((r) => Number((r.meta && r.meta.changes) ?? 0));
  if (inserted.some((c) => c === 0)) {
    const mine = nulls.filter((_, i) => inserted[i] === 1).map((n) => n.nul);
    if (mine.length) await env.CREDITS_DB.batch(mine.map((nul) => env.CREDITS_DB.prepare("DELETE FROM spent WHERE nullifier = ?1 AND ts = ?2").bind(nul, ts)));
    return { ok: false, why: "credit token already spent" };
  }
  await env.CREDITS_DB.prepare("INSERT INTO stats (kid, issued, redeemed) VALUES (?1, 0, ?2) ON CONFLICT(kid) DO UPDATE SET redeemed = redeemed + ?2").bind(pool.issuing.kid, cost).run();
  return { ok: true, why: "" };
}

export function creditCost(method, path) {
  return CREDIT_COSTS[`${method.toUpperCase()} ${path}`] ?? null;
}
